### ISSUE TITLE
ENH: LinearOperator for linprog arguments A_ub and A_eq

### ISSUE THREAD

**Issue by lorentzenchr** [scipy/scipy#15185]
### Is your feature request related to a problem? Please describe.

The estimation of linear quantile regression can be recast as a linear programming problem with equality matrix
```python
n_samples = X.shape[0]
A_eq = np.concatenate(
                [X, -X, np.eye(n_samples), -np.eye(n_samples)], axis=1
            )
```
where `X` is the feature matrix (predict will give `X @ coef`), see https://github.com/scikit-learn/scikit-learn/blob/0d378913be6d7e485b792ea36e9268be31ed52d0/sklearn/linear_model/_quantile.py#L203-L217.

This could help to save computation time and memory.

### Describe the solution you'd like.

I'd like be able to pass a `scipy.sparse.linalg.LinearOperator` to `linprog` as argument `A_ub` and `A_eq`.

**Comment by ilayn** [comment]
Do you mean sparse.linalg.LinearOperator? Also what would be the cost saving with a LinearOperator instance and what is saved?

**Comment by mdhaber** [comment]
@mckib2 Do you think we can do this with HiGHS? If not with the existing capabilities, is it something we should ask about?

**Comment by jajhall** [comment]
No, HiGHS cannot handle anything other than an actual sparse matrix. 

It's conceivable that the simplex solver could be adapted to handle an abstracted linear operator as a means of obtaining columns of the implied constraint matrix, and performing matrix-vector products. However, a square submatrix of the constraint matrix would have to be formed periodically as an actual sparse matrix, so some overhead comparable to supplying the whole matrix in that form would be inevitable unless the whole matrix has overwhelmingly more columns than rows.

Although it's an interesting challenge, incorporating it would be a lot of work, for which resources are currently unavailable. Thus, whoever wanted it would have to come up with a lot of cash!

**Comment by mdhaber** [comment]
Thanks @jajhall.

@lorentzenchr Please see @ilayn's [questions above](https://github.com/scipy/scipy/issues/15185#issuecomment-989307278). Your answer will help us decide whether this issue should be left open or closed. It's something we'd need to consider adding to the roadmap, as it's not something that would be quick/easy to implement (efficiently).

**Comment by lorentzenchr** [comment]
> Also what would be the cost saving with a LinearOperator instance and what is saved?

I think the main advantage would be to avoid twice the storage of the feature matrix `X`.

Thinking about it twice (or some power of 2), maybe the best thing to do is using an orthogonal transformation of the decision variable `x`: The current `x` contains slack variables `s, t > 0` for the coefficients of the linear model, i.e. `coef = s - t`. Instead of working with `s` and `t` one could work with `s + t >= 0` and (unbounded) `s - t`. I think the transformed equality matrix would become
```python
A_eq = np.concatenate([X, np.zeros_like(X), np.eye(n_samples), -np.eye(n_samples)], axis=1)
```
This is much better for memory (and computation).
Are there any downsides in doing so? Is the problem internally transformed back into positive slack variables?

**Comment by jajhall** [comment]
There are two issues here:
1. Any simplex solver will be much more efficient using two explicit copies of X than working with some fancy operator. So, there's no performance gain to be had. 
2. It's hard for me to reverse-engineer the code into the precise mathematical formulation of the problem, but is your copy of X is due to the coefficients of the linear model being "free" variables? Are s>0 and t>0 used to express your free variables coeff as s-t, in terms of non-negative variables? If so, then it's unnecessary: just have coeff and make the corresponding bounds -Inf and +Inf. This may well be what your "orthogonal transformation" is, any way.
3. Internally, HiGHS uses generally bounded variables and generally bounded constraints, with any of the bounds being infinite if necessary. There's no distinction between inequalities and equations, and constraints can have finite lower and upper bounds - ie "boxed" - like variables. There are not even explicit slacks.

**Comment by lorentzenchr** [comment]
Thank you all for you super fast and insightful replies. I guess, with @jajhall last explanations, the better option is indeed to use rotated variables.

For the interested reader, here is the derivation of [linear quantile regression](https://scikit-learn.org/stable/modules/linear_model.html#quantile-regression) into a linear programming problem, see also https://stats.stackexchange.com/questions/384909/:
#### Linear Quantile Regression
Given data `X` and `y`, estimate the quantile at level `q` as `X @ coef + intercept`. One usually uses the pinball loss and we add an L1 penalty of strength `alpha` for `coef`:
```
argmin_{coef, intercept} 1/n sum_i pinball_loss(y_i, x_i @ coef + intercept) + alpha ||coef||_1
```
#### Use linear programming formulation of quantile regression
```
min_x c x
      A_eq x = b_eq
      0 <= x
```
with the definitions
```
x = (s0, s, t0, t, u, v) = slack variables
intercept = s0 - t0
coef = s - t
c = (0, n * alpha * 1_p, 0, n * alpha * 1_p, q * 1_n, (1-q) * 1_n)
residual = y - X@coef - intercept = u - v
A_eq = (1_n, X, -1_n, -X, diag(1_n), -diag(1_n))
b_eq = y
p = n_features
n = n_samples
1_n = vector of length n with entries equal one
```
With (back-) rotation of `s` and `t` to
```
z = (s0 + t0, s + t, s0 - t0, s - t, u, v)
d = (0, n * alpha * 1_p, 0, q * 1_n, (1-q) * 1_n)
B_eq = (0, 0_p, 1_n, X, diag(1_n), -diag(1_n))
```
the problem reads
```
min_z d z
      B_eq x = b_eq
      0 <= z[0:p+1]
      -inf < z[p+1:] < inf
```
The important point is that `B_eq` has the first p+1 columns all zeros and the matrix `X` only enters once.

**Comment by jajhall** [comment]
Thanks for this. However, the components of z corresponding to u and v are non-negative, not free as you suggest.

More seriously, as I understand it, the reformulation doesn't correspond to the original problem. 
- Let z2 denote your second set of components of z
- Let z4 denote your fourth set of components of z
Now, z2 have positive objective coefficients (n*alpha), don't appear in the equations, and are non-negative. Hence their optimal value in the LP is z2=0. Given what z2 represents, this means that the 1-norm of the coefficients is zero. Indeed, since z2=s+t, with s>=0 and t>=0, the corresponding values of s and t can only be zero. This will conflict with the optimal value of z4, which is s-t, and can't be expected to be zero.

For me the correct reformulation retains s0, s, t0, t, u, v, and introduces a variable z (nothing to do with your z, but I can't think of a better letter!) that is the value of the coefficients. The term Xs - Xt is replaced by Xz (so X no longer appears twice) and the constraint z=s-t is introduced. 

It's unfortunate that the "community" thinks that "Linear programs are predominantly analyzed and solved using the standard form min c^Tz subject to Az=b, z>=0". Analysed, yes, but not solved. I guess that this leads people to want to formulate problems as LPs in this form. 

Looking at your original problem, I'd not worry about coeff (ie w) being "free" variables. I'd recognise that ||w||_1 is only piecewise linear, so a standard LP solver won't handle it, and introduce variables s and t for the positive and negative parts of w. Thus, after adding the constraint w=s-t, ||w||_1 can be replaced by 1^T(s+t). This leads directly to my correction of your formulation.

**Comment by jajhall** [comment]
Slightly off-topic, and if you're a computational statistician, I guess that you've used R. We were thinking of creating a R interface to HiGHS. Do you think that this would be worthwhile?

**Comment by lorentzenchr** [comment]
@jajhall Thanks for looking into it again. I added the bounds on u and v.

IIUC, you are suggesting to use and solve for `x = (intercept, coef, s, t, u, v)` with constraints:
- `y = X@coef + intercept + u - v` (as before)
- `coef = s - t`. This will lead to p+1 new rows in `A_eq`.

> It's unfortunate that the "community" thinks that "Linear programs are predominantly analyzed and solved using the standard form min c^Tz subject to Az=b, z>=0". Analysed, yes, but not solved. I guess that this leads people to want to formulate problems as LPs in this form.

Documentation and examples would help in this case.

> Slightly off-topic, and if you're a computational statistician, I guess that you've used R. We were thinking of creating a R interface to HiGHS. Do you think that this would be worthwhile?

While R has the excellent [quantreg](https://CRAN.R-project.org/package=quantreg) package, I think it would still make sense to have this high quality solver available in the R-verse.

**Comment by jajhall** [comment]
> @jajhall Thanks for looking into it again. I added the bounds on u and v.
> 
> IIUC, you are suggesting to use and solve for `x = (intercept, coef, s, t, u, v)` with constraints:
> 
> * `y = X@coef + intercept + u - v` (as before)
> * `coef = s - t`. This will lead to p+1 new rows in `A_eq`.

Yes, there will be p+1 new rows for the `coef = s - t` constraints

> 
> > It's unfortunate that the "community" thinks that "Linear programs are predominantly analyzed and solved using the standard form min c^Tz subject to Az=b, z>=0". Analysed, yes, but not solved. I guess that this leads people to want to formulate problems as LPs in this form.
> 
> Documentation and examples would help in this case.

Where would the documentation and examples be? I could link a short course on LP to the HiGHS website, but I feel that people are drawn to a solver after they've formulated their problems.

> 
> > Slightly off-topic, and if you're a computational statistician, I guess that you've used R. We were thinking of creating a R interface to HiGHS. Do you think that this would be worthwhile?
> 
> While R has the excellent [quantreg](https://CRAN.R-project.org/package=quantreg) package, I think it would still make sense to have this high quality solver available in the R-verse.

Thanks. I see that quantreg solves the LPs using interior point method (IPM) solvers. They have trouble with free variables.

It would be interesting to benchmark their IPM solver against ours. I assume that the matrix X is sparse. They also talk about the need to solve parametric linear programming problems for one variant, and that his is "very slow".  I assume that the matrix X is sparse. I'll try contacting folk involved with this project to sound out the strength of the LP software available using R

**Comment by lorentzenchr** [comment]
> Where would the documentation and examples be?

In the [scipy docu of linprog](https://scipy.github.io/devdocs/reference/generated/scipy.optimize.linprog.html#scipy-optimize-linprog). Something like: "There is no need to formulate a linear programming problem in terms of positive variables. The solvers accept general bounds for `x` and might even be more efficient this way."

**Comment by lorentzenchr** [comment]
Shall we close this issue? The highs*-methods really digest sparse (csc) matrices and the other solvers are...maybe....not that interesting.

**Comment by jajhall** [comment]
> > Where would the documentation and examples be?
> 
> In the [scipy docu of linprog](https://scipy.github.io/devdocs/reference/generated/scipy.optimize.linprog.html#scipy-optimize-linprog). Something like: "There is no need to formulate a linear programming problem in terms of positive variables. The solvers accept general bounds for `x` and might even be more efficient this way."
> 
> A further doc improvement could be to say, that highs as well as interior-point, in the end, work on csc matrices.

I see what you mean now. Thanks. Been a productive discussion for me, but I think it's run its course. Cheers Julian

**Comment by mdhaber** [comment]
Shall we close this issue? 

Done. Thanks, all.
